janusgraph-core/src/main/java/org/janusgraph/graphdb/database/idhandling/VariableString.java

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.database.idhandling;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang.StringUtils;
+import org.janusgraph.diskstorage.ReadBuffer;
+import org.janusgraph.diskstorage.WriteBuffer;
+import org.janusgraph.graphdb.database.serialize.attribute.StringSerializer;
+import org.janusgraph.util.encoding.StringEncoding;
+
+import static org.janusgraph.graphdb.database.idhandling.VariableLong.STOP_MASK;
+
+
+/**
+ * Handle String serialization and deserialization, support both forward and backward read and write.
+ * This does not use any compression technique, so it is most suitable for short string, e.g. vertex
+ * id.
+ *
+ * Important: it uses {@link VariableLong#STOP_MASK} as the first byte. This is because JanusGraph used
+ * to assume all IDs are positive long values. To maintain backward compatibility, we have to make sure
+ * JanusGraph knows how to distinguish between string ID and long ID when decoding binary data. The
+ * convention is, if the first byte is STOP_MASK, then it's a string ID, otherwise it's a long ID. This
+ * leverages the fact that IDs are always positive, so the first byte would never be STOP_MASK for long
+ * IDs.
+ *
+ * This class uses {@link StringSerializer} as a reference.
+ *
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
```

Comment:
Is it the byte which tells that the key is String or Long? (https://github.com/JanusGraph/janusgraph/pull/3418#discussion_r1171987782)
If so, I just wander why we need it exactly. Unless users want to use a combination of keys (i.e. `Long` for some vertices and `String` for other vertices), this marker won't be useful that much unless I'm missing something.
Is it here so that a single JanusGraph instance could work with both ID types simultaneously? 
I.e. JanusGraph instance already knows weather it works with String ids or Long ids. Not sure we need to check the key type for each key we retrieve. 
I definitely not against using this byte, I just not sure I understand why we need to use it.

Reply:
> Is it the byte which tells that the key is String or Long? (https://github.com/JanusGraph/janusgraph/pull/3418#discussion_r1171987782)

Yes.

>  Unless users want to use a combination of keys (i.e. Long for some vertices and String for other vertices), this marker won't be useful that much unless I'm missing something

To some extent, you are right. This marker is helpful in:
I. Allow users to use `Long` for some vertices and `String` for other vertices.
II. (same as I., but from a different perspective) Allow existing JanusGraph users to start using string-type IDs without creating a new graph from scratch (a.k.a. ability to "upgrade"). This opens a door to those who would like to "gradually" abandon long-type IDs.
III. (same as I., but from a different perspective) Allow users to stop using string-type IDs without creating a new graph from scratch (a.k.a. ability to "downgrade").
IV. Make the codebase cleaner. Otherwise, we would need a boolean/enum variable across the codebase to control whether JanusGraph is dealing with long-type IDs or string-type IDs.

So who benefits from this design and who suffers from it?
- Users who stick to long-type IDs: Neither overhead nor benefit for them.
- Existing users who would like to start using string-type IDs: This is a must-have feature for them.
- New users who are only interested in using string-type IDs: One byte overhead for ID :(
- Developers: minimal benefit as code is cleaner this way.

Reply:
Actually, the most important reason is we MUST do it (not 100% sure, though). Schema definitions are modeled as vertices and edges in JanusGraph, too. Since these vertices are of long-type, there is currently no way of getting rid of long-type IDs completely.

Reply:
I do agree that this byte give users ability to slowly migrate to String id or to migrate back to Long, but I think there are some additional cons we need to consider:
- Additional runtime branching logic. Even so JIT might optimize many of those branches they will give some overhead (probably negligible, but still). In case a single JanusGraph instance support only a single type of ID then we could make refactoring and configure JanusGraph to use only the specific ID strategy.
- Additional logic to understand for a developer. In case of a separate strategies abstracted away by a common interface, the developer won't need to read and understand code for both String IDs and Long IDs. Instead, it's enough to read a single strategy which developer is interested in.
- In case we add any other ID types the byte might be not very useful as it is designed specifically for String type. I.e. it basically means that any 9 bytes array with a STOP_MASK is an 8 bytes array. It won't help if we want to have 16 bytes String and UUID type or byte array type as ID type. 

> Actually, the most important reason is we have to do it (not 100% sure, though). Schema definitions are modeled as vertices and edges in JanusGraph, too. Since these vertices are long-type, there is currently no way of getting rid of long-type IDs completely. That being said, it's possible to add special control options in the codebase so that we know that we are reading normal vertices or schema vertices (assuming we would like to get rid of the marker byte).

Agree that we have to keep logic for Long ids always due to Schema vertices. This does seems to be a use case when we need to use `Long` type always and `another type` optionally.
In such case, I think we could leave this byte as is, but I would say that it makes sense to think about this byte not as `String` marker byte but as `non-default type` marker byte.
Meaning that this marker means that the type of this key is equal to the one, configured in JanusGraph instance (could be String, UUID, byte array, or something else serializable).
That said, in such case, we need to always store this extra byte for any byte array grater then 7 bytes. Otherwise it won't be possible to tell weather this 9 bytes array is actually of size 9 or 8. Currently you store this byte only for a String with length == 8, but it's possible just because you use ASCII characters in the String and non-ASCII character for a marker. For byte arrays it won't be a good idea to restrict usage of the first byte.
What do you think about it? Do you think it makes sense to store this byte only for String.length == 8 (improves efficiency of Strings which are bigger then 8 bytes) or store this byte for String.length >=8 (adds possibility of implementing other data types later, but sacrifices 1 byte for each type >= 8 bytes)?

Reply:
Just to be clear, my current approach uses two different encoding schemes for STRING:
1. In `VariableString` class, it uses a special STRING marker byte at the beginning.
2. In `IDManager::getKeyID` class, it relies on the fact that the length of STRING id is not 8 bytes (we add one dummy byte if its length is 8).

Note that the first scenario applies to encoding vertex IDs in edges and indexes, whereas the latter scenario applies to encoding vertex IDs as partition keys.

For the latter case, you brought up a good point: if we use `length == 8` to judge whether it's LONG or STRING, it leaves no room for any future extension where we might allow byte array as ID. In that case, I think we should always add a marker at the beginning when it's not long-type.

Reply:
Updated `IDManager::getKeyID`. Now if JanusGraph sees the buffer has 8 bytes, it knows it's a long-type ID. Otherwise:
1) If the first byte is `STRING_ID_MARKER`, then it's a long ASCII string. The end of the string is determined by the `STOP_MASK`. (We add a dummy padding byte if applicable, to make sure the byte buffer is always not 8 bytes).
2) Otherwise, it's neither long-type nor string-type. This leaves room for future extensions, e.g. byte array ID.